simulator/func_sim/func_sim.cpp

```diff
@@ -58,23 +58,54 @@ typename FuncSim<ISA>::FuncInstr FuncSim<ISA>::step()
 template <typename ISA>
 void FuncSim<ISA>::init( const std::string& tr)
 {
-    mem->load_elf_file( tr);
-    PC = mem->startPC();
-    nops_in_a_row = 0;
+    load_binary_file( tr);
+    prepare_to_run();
 }
 
 template <typename ISA>
-void FuncSim<ISA>::run( const std::string& tr, uint64 instrs_to_run)
+void FuncSim<ISA>::run( uint64 instrs_to_run)
 {
-    init( tr);
-    for ( uint32 i = 0; i < instrs_to_run; ++i) {
+    for ( uint64 i = 0; i < instrs_to_run; ++i) {
         const auto& instr = step();
         sout << instr << std::endl;
         if ( instr.is_halt())
             break;
     }
 }
 
+template <typename ISA>
+void FuncSim<ISA>::load_binary_file( const std::string &tr) {
+    mem->load_elf_file( tr);
+}
+
+template <typename ISA>
+void FuncSim<ISA>::prepare_to_run() {
+    PC = mem->startPC();
+    nops_in_a_row = 0;
+}
+
+template <typename ISA>
+size_t FuncSim<ISA>::mem_read (Addr addr, unsigned char *buf, size_t length) try {
```

Comment:
Is there any problem with pushing to FuncMemory class?

Reply:
`FuncMemory::read` is non-static, so it must be called for an object. Memory members are defined in `FuncSim` and `PerfSim` classes, but not in generic `Simulator`.  
If I'll make it static, then I'll have to make `FuncMemory::memory` static and we won't be able to have more than one simulator instance.

Reply:
Right, but I mean moving the code itself. The final result should look like this:

```c++
template<typename ISA>
size_t FuncSim<ISA>::mem_read(Byte* dst, Addr src, size_t length) const {
    return mem->memcpy_guest_to_host(dst, src, length);
}

template<typename ISA>
size_t FuncSim<ISA>::mem_write(Addr addr, const Byte* buf, size_t length) {
    return mem->memcpy_host_to_guest(dst, src, length);
}
```

Reply:
Additionally, they may be wrapped to the exception-less wrappers:

```c++
template<typename ISA>
size_t FuncSim<ISA>::mem_read_noexcept(Byte* dst, Addr src, size_t length) const noexcept try {
    return mem_read(dst, src, length);
}
catch (...) {
    return 0;
}
```

Reply:
This looks good for reading memory. What about writing to memory? I think this would require `InstrMemory` method to erase cache from `src` to `src + length` before writing, so it would look like:
```
size_t FuncSim<ISA>::mem_read(Byte* dst, Addr src, size_t length) const {
    mem->clear_cache(src, length);
    return mem->memcpy_guest_to_host(dst, src, length);
}

Reply:
Right. I added similar method for writing a single chunk of data.